apollo-federation/src/sources/connect/json_selection/parser.rs

```diff
@@ -21,12 +22,53 @@ use super::helpers::spaces_or_comments;
 use super::known_var::KnownVariable;
 use super::lit_expr::LitExpr;
 use super::location::merge_ranges;
+use super::location::new_span;
 use super::location::ranged_span;
 use super::location::OffsetRange;
 use super::location::Ranged;
 use super::location::Span;
 use super::location::WithRange;
 
+// ParseResult is the internal type returned by most ::parse methods, as it is
+// convenient to use with nom's combinators. The top-level JSONSelection::parse
+// method returns a slightly different IResult type that hides implementation
+// details of the nom-specific types.
+//
+// TODO Consider switching the third IResult type parameter to VerboseError
+// here, if error messages can be improved with additional context.
+pub(super) type ParseResult<'a, T> = IResult<Span<'a>, T>;
```

Comment:
In the future (not suggesting you need to go and change the ones in this PR), if you use triple-slash quotes right before an item, it sets that comment as the docs for the item. Docs also support Markdown syntax, so you can get nicely-rendered docs for things in your editor. Here's in RustRover, where that link to `JSONSelection::parse` actually brings you to the right code!

![Screenshot 2024-11-01 at 10 59 47 AM](https://github.com/user-attachments/assets/3a75703b-4c94-43b7-9240-850f02debd1f)
![Screenshot 2024-11-01 at 11 00 41 AM](https://github.com/user-attachments/assets/7686753c-7c2a-44f6-b180-d653561fa3c2)